### PR TITLE
Remove ExpressionLanguage dependency by using callbacks for the `where` modifier

### PR DESCRIPTION
**Pull request by alexandre-daubois**

### CODE REVIEW

File: src/Query/Modifier/Where.php [hunk 1]
```diff
@@ -4,37 +4,28 @@
...
         $final = [];
         foreach ($source as $item) {
-            $localContext = [$this->parentQuery->getSourceAlias() => $item] + $this->environment;
-            $localContext += $context->getEnvironment($item);
+            $localContext = $context
+                ->withEnvironment($item, [$this->parentQuery->getSourceAlias() => $item])
+                ->getEnvironment($item);
 
-            if ($this->expressionLanguage->evaluate($this->expression, $localContext)) {
+            if (true === $this->callback->call($this, $item, new QueryContextEnvironment($localContext))) {
```
Comment (Jean-Beru) [inline src/Query/Modifier/Where.php:28]: Is it needed to use `$this` context ? Callback will be executed in a `Where` context.

File: src/Query/Modifier/Where.php [hunk 2]
```diff
@@ -4,37 +4,28 @@
...
         parent::__construct($parentQuery);
 
-        $this->expression = new Expression($expression);
-        $this->environment = $environment;
-        $this->expressionLanguage = new ExpressionLanguage();
-
-        foreach (Query::getRegisteredWhereFunctions() as $function) {
-            $this->expressionLanguage->addFunction($function);
-        }
+        $this->callback = $callback(...);
```
Comment (Jean-Beru) [inline src/Query/Modifier/Where.php:17]:
```suggestion
        $this->callback = Closure::fromCallable($callback)->bindTo(null);
```
?

Comment (alexandre-daubois) [reply src/Query/Modifier/Where.php:17]: It seems it is equivalent! PHPStorm tells me that `'Closure::fromCallable()' can be converted to the first-class callable syntax `

File: src/Query/Modifier/Where.php (hunk 1 above)
Comment (alexandre-daubois) [reply src/Query/Modifier/Where.php:28]: I must set a `$newThis` to call, it cannot be null. Maybe I miss something?

**Pull request merged**